management/src/main/java/io/micronaut/management/endpoint/management/ManagementController.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2017-2021 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.management.endpoint.management;
+
+import io.micronaut.context.annotation.Requires;
+import io.micronaut.http.HttpRequest;
+import io.micronaut.http.annotation.Controller;
+import io.micronaut.http.annotation.Get;
+import io.micronaut.http.server.util.HttpHostResolver;
+import io.micronaut.runtime.server.EmbeddedServer;
+import io.micronaut.web.router.UriRoute;
+import io.reactivex.Single;
+
+import java.util.stream.Stream;
+
+/**
+ * <p>Exposes the available management endpoints of the application.</p>
+ *
+ * @author Hernán Cervera
+ * @since 3.0.0
+ */
+@Controller
+@Requires(beans = EmbeddedServer.class)
+public final class ManagementController {
+
+    private final HttpHostResolver httpHostResolver;
+
+    private final ManagementRoutesResolver managementRoutesResolver;
+    private final ManagementDataCollector<?> managementDataCollector;
+
+    public ManagementController(HttpHostResolver httpHostResolver,
+                                ManagementRoutesResolver managementRoutesResolver,
+                                ManagementDataCollector<?> managementDataCollector) {
+        this.httpHostResolver = httpHostResolver;
+        this.managementRoutesResolver = managementRoutesResolver;
+        this.managementDataCollector = managementDataCollector;
+    }
+
+    @Get("management")
+    public Single<?> getManagementRoutes(HttpRequest<?> httpRequest) {
+        String routeBase = httpHostResolver.resolve(httpRequest);
+        String managementDiscoveryPath = httpRequest.getPath();
+        Stream<UriRoute> uriRoutes = managementRoutesResolver.getRoutes();
```

Comment:
* Replace `Single` with `Mono` and don't use RxJava 2 but instead use Reactor
* I believe this should also be an `@Endpoint` to take advantage of the security / port binding features
* I don't think the current path or route base needs resolving, this can be discovered from https://github.com/micronaut-projects/micronaut-core/blob/f2e1d122efc4ee7bb2e7249c20dc1c4f2ebccdf1/management/src/main/java/io/micronaut/management/endpoint/EndpointDefaultConfiguration.java and it needs to factor in the path and the port

Reply:
This is a `@Controller` because in [this comment](https://github.com/micronaut-projects/micronaut-core/pull/5713#issuecomment-875735950) James suggested to me that a `@Controller` would be more appropriate to discover HTTP links (since [the author of the issue was hoping for clickable URLs](https://github.com/micronaut-projects/micronaut-core/issues/5006#issuecomment-786981171)); something of the following format:

```text
http://localhost:8080/health
```

Building such a URL requires knowing: protocol, host and port.

I have learned that the proper way to discover this data is through the `HttpHostResolver`, but it doesn't seem it can be used in an `@Endpoint` since there is no access to the request.

I think that making this an `@Endpoint` and using `EndpointDefaultConfiguration` as you suggest makes it possible to create links of the following format, wich are not clickable:

```text
:9090/health    # When the port is configured through "endpoints.all.port".
/health         # When the port is the same as the application's port.
```

Would you approve the previous link format, or should it be clickable (contain protocol, host, port)? If you think the latter is required, I would appreciate some pointers into how to discover the app's protocol and port using an `@Endpoint`

Reply:
Yeah I am unconvinced this should be `@Controller` since I would expect all of the configuration under `EndpointDefaultConfiguration` to apply to this endpoint as well (the port it is exposed under, the sensitivity, the path exposed under) and as written none of this is respected. Maybe @jameskleeh can elaborate more what he had in mind with making it a `@Controller`

Reply:
@jameskleeh Right, but my question remains, how do you get this new controller to respect the port management is exposed under, the sensitivity and the path management is exposed under.

Right now this would expose this information under '/management` with seemingly no way to configure that and it doesn't respect `endpoints.all.port` so you can't send it to another port either.

As it written it is a breaking change if anyone has an existing URI mapped to `/management`, plus many users run the management endpoints on a different port to the actual application for security reasons.

Reply:
@graemerocher I don't think the requirement exists for it to respect that configuration because it's not an endpoint. I'm in support for the URI to be configurable and be able to be turned off.

I don't consider a new feature that maps to a given endpoint to be a breaking change, otherwise we wouldn't be able to add many features. For example we couldn't add a new exception handler because one may already exist for that exception. The security module couldn't add any endpoints because it may already be mapped to another controller (this example). We wouldn't be able to add new type converters, etc... the list goes on.

Reply:
I don't think it makes any sense whatsoever to have this behave differently and is a potential security vulnerability if you expose the port where the management endpoints do exist over the default port.

Spring Boot treats this as part of the endpoints and respects the path / port configuration (the equivalent of `endpoints.all.port` and `endpoints.all.path` https://docs.spring.io/spring-boot/docs/current/reference/html/actuator.html#actuator.endpoints.hypermedia

I don't also buy the argument that in the context of JMX there is no URI because in the context of JMX `endpoints.all.path` is not relevant either.

Reply:
> if you expose the port where the management endpoints do exist over the default port

The management endpoints are exposed over the default port afaik. I think this discovery api should create links with the configured endpoint port though.

I don't think it's relevant what spring boot does.

The management module supports the notion of a configurable base URI because we expose the endpoints over the web by default. So correct its not relevant to JMX and this new "endpoint" isn't either. I'm not sure the point you're trying to make here.

Reply:
Spring has the notion of `@WebEndpoint` and endpoints specific to the web. I don't understand the point of doing so given the main point of endpoints is to not be web specific. Perhaps the thing to do would be to understand why they decided to do that and then determine if something similar makes sense here.

Reply:
> The management endpoints are exposed over the default port afaik.

They are not if you specify `endpoints.all.port` to expose them over a different port and you can customize the path with `endpoints.all.path`

As a very least if `endpoints.all.path` is specified the controller should map to that and if `endpoints.all.port` is specified then `Controller(port)` for this controller should map to that